lib/llm/src/kv_router/scheduler.rs

```
@@ -51,166 +37,159 @@ pub enum KvSchedulerError {
 
-        // Get worker IDs from instances
-        let worker_ids: Vec<i64> = instances.iter().map(|i| i.instance_id).collect();
-
         let (event_tx, event_rx) = tokio::sync::mpsc::unbounded_channel::<KVHitRateEvent>();
+        let ns_clone = component.namespace().clone();
         tokio::spawn(async move {
             let mut event_rx = event_rx;
             while let Some(event) = event_rx.recv().await {
-                if let Err(e) = ns.publish(KV_HIT_RATE_SUBJECT, &event).await {
+                if let Err(e) = ns_clone.publish(KV_HIT_RATE_SUBJECT, &event).await {
                     tracing::warn!("Failed to publish KV hit rate event: {:?}", e);
                 }
             }
         });
 
-        let sequences = Arc::new(Mutex::new(ActiveSequencesMultiWorker::new(
+        let worker_ids: Vec<i64> = instances
+            .iter()
+            .map(|instance| instance.instance_id)
+            .collect();
+        let slots = Arc::new(ActiveSequencesMultiWorker::new(
+            component,
             block_size as usize,
             worker_ids,
-        )));
+            replica_sync,
+        ));
 
-        // Channel to accept new scheduling requests
+        let slots_clone = slots.clone();
         let (request_tx, request_rx) = tokio::sync::mpsc::channel::<SchedulingRequest>(1024);
         // Background task to handle scheduling requests
         tokio::spawn(async move {
-            let mut request: SchedulingRequest;
             let mut request_rx = request_rx;
-            let mut pending_endpoint_update: Option<Vec<i64>> = None;
             tracing::trace!("scheduler background task started");
 
-            'outer: loop {
-                request = tokio::select! {
-                    biased;
+            loop {
+                // First, check for instance updates (non-blocking)
+                if instances_rx.has_changed().unwrap_or(false) {
```

If has_changed returns an `Err`, we may want some sort error handling. I don't think we should be silently failing here.